### ISSUE TITLE
Untoward universe problem

### ISSUE THREAD

**Issue by StevenTschantz** [coq/coq#16068]
I ran across a problem I've traced down to some sort of universe problem.  Something defaults one way in one context but changes unexpectedly in another.  I've reduced my problem to a simple schematic form illustrated below.  The testFgood passes, but the testFbad fails.  Apparently the "if" return type defaults to a different universe in the context of being in the return of the match.

I'm running 8.13.2 on a Windows machine, so I am a little behind.

```coq
Inductive testI : bool -> Type :=
| testItrue : testI true
| testIfalse : testI false
.

Definition testFgood (b : bool) (x : testI b) :
if b then b = b else b = b.
Proof.
refine (
match x as x0 in testI b0 return
  if b0 return Prop then b0 = b0 else b0 = b0
with
| testItrue => eq_refl
| testIfalse => eq_refl
end
).
Qed.

Definition testFbad (b : bool) (x : testI b) :
if b then b = b else b = b.
Proof.
refine (
match x as x0 in testI b0 return
  if b0 then b0 = b0 else b0 = b0
with
| testItrue => eq_refl
| testIfalse => eq_refl
end
).
Qed.
```

- Steven Tschantz

**Comment by Alizter** [comment]
If you use
```coq
match x as x0 in testI b0 return
  if b0 return Prop then b0 = b0 else b0 = b0
with
| testItrue => eq_refl
| testIfalse => eq_refl
end
```
Then it works. It is not a universe problem, but that match thinks it is living in Set(/Type) for some reason, rather than in Prop.

**Comment by Alizter** [comment]
If we have a look at your term:
```coq
match
       x in (testI b0)
       return
         match b0 return Prop with
         | true => @eq bool b0 b0
         | false => @eq bool b0 b0
         end
     with
     | testItrue => @eq_refl bool true
     | testIfalse => @eq_refl bool false
     end 
```

We see that the return type of the outer match is not `Prop` until `b0` has been resolved. However it does sit in `Type` therefore Coq defaults to that.

**Comment by StevenTschantz** [comment]
I agree this is what is happening, but not that I wanted it to happen, or could easily understand it when it happened unexpectedly.  Thus I termed this "untoward" though unexpected and hard to diagnose also apply.

So how does the universe return type of match b0 ... end depend on the resolution of the type of b0 when the branches are @eq bool _ _ that surely must live in Prop?  Is there some default rule that the universe of a match returning in Prop/Set/Type is the max of the returns and the matched, or was this a problem with when the type of the inner match was resolved?

I understand determining the type returned from a match isn't so simple and sometimes I need to make the return type explicit, but this was, again, unexpected.  Thanks for your quick reply and formatting.

**Comment by SkySkimmer** [comment]
similar issue can be seen if we do
~~~coq
Set Printing All.
Check match true as b return if b then True else True with true => I | false => I end.
(* match true as b return match b return Set ... *)
~~~
the return type when provided is checked against a fresh sort so it's like `(if b then True else True) : Type@{_}` (but with the cast disappearing) which indeed also produces `return Set`.

Code is https://github.com/coq/coq/blob/3703953ce9bd6e850ca81381e2827e102c1f5120/pretyping/cases.ml#L2156-L2157 dating from apparently 066998ff9f89404cb41a6a3f8c0ccc88d2cc7a4b
To avoid passing `typing_fun` a sort (if we pass a sort we have already lost) we need some other way so ensure it produces a type (and we need to be compatible with coercions to sortclass)

**Comment by SkySkimmer** [comment]
Fixed in 8.18 (probably https://github.com/coq/coq/pull/16903 or a followup)
